ext/backtracie_native_extension/backtracie_frames.c

```diff
@@ -0,0 +1,787 @@
+    loc->self_is_real_self = 1;
+  } else {
+    loc->self_or_self_class = rb_class_of(cfp->self);
+    loc->self_is_real_self = 0;
+  }
+  loc->pc = cfp->pc;
+  return true;
+}
+
+bool backtracie_capture_frame_for_thread(VALUE thread, int frame_index,
+                                         raw_location *loc) {
+  if (!is_thread_alive(thread))
+    return 0;
+  rb_thread_t *thread_pointer = (rb_thread_t *)DATA_PTR(thread);
+
+#ifndef PRE_EXECUTION_CONTEXT
+  return backtracie_capture_frame_for_execution_context(thread_pointer->ec,
+                                                        frame_index, loc);
+#else
+  return backtracie_capture_frame_for_execution_context(thread_pointer,
+                                                        frame_index, loc);
+#endif
+}
+
+bool backtracie_capture_frame(int frame_index, raw_location *loc) {
+#ifndef PRE_EXECUTION_CONTEXT
+  return backtracie_capture_frame_for_execution_context(GET_EC(), frame_index,
+                                                        loc);
+#else
+  // FIXME: Figure out how to make GET_EC (GET_THREAD) work for Ruby <= 2.4
+  return false;
+#endif
+}
+
+int backtracie_frame_line_number(const raw_location *loc, size_t prev_loc_len) {
+  const raw_location *next_rbframe = next_ruby_location(loc, prev_loc_len);
+  if (next_rbframe) {
+    return calc_lineno((rb_iseq_t *)next_rbframe->iseq,
+                       (const VALUE *)next_rbframe->pc);
+  } else {
+    return 0;
+  }
+}
+
+size_t backtracie_frame_name_cstr(const raw_location *loc, char *buf,
+                                  size_t buflen) {
+  strbuilder_t builder;
+  strbuilder_init(&builder, buf, buflen);
+
+  method_qualifier(loc, &builder);
+  method_name(loc, &builder);
+
+  return builder.attempted_size;
+}
+
+VALUE backtracie_frame_name_rbstr(const raw_location *loc) {
+  strbuilder_t builder;
+  strbuilder_init_growable(&builder, 256);
+
+  method_qualifier(loc, &builder);
+  method_name(loc, &builder);
+
+  VALUE ret = strbuilder_to_value(&builder);
+  strbuilder_free_growable(&builder);
+  return ret;
+}
+
+size_t backtracie_frame_filename_cstr(const raw_location *loc,
+                                      size_t prev_loc_len, bool absolute,
+                                      char *buf, size_t buflen) {
+  strbuilder_t builder;
+  strbuilder_init(&builder, buf, buflen);
+
+  frame_filename(loc, prev_loc_len, absolute, &builder);
+
+  return builder.attempted_size;
+}
+
+VALUE backtracie_frame_filename_rbstr(const raw_location *loc,
+                                      size_t prev_loc_len, bool absolute) {
+  strbuilder_t builder;
+  strbuilder_init_growable(&builder, 256);
+
+  int fname_found = frame_filename(loc, prev_loc_len, absolute, &builder);
+
+  VALUE ret = Qnil;
+  if (fname_found) {
+    ret = strbuilder_to_value(&builder);
+  }
+  strbuilder_free_growable(&builder);
+  return ret;
+}
+
+static int frame_filename(const raw_location *loc, size_t prev_loc_len,
```

Comment:
Would it make sense to make the return type a `bool` here?

Reply:
Yeah I haven't been super consistent about that. I'll do a pass and turn all the ints that should be bools into bools.